src/aggregation/agg_req_with_accessor.rs

```diff
@@ -95,33 +99,37 @@ impl AggregationWithAccessor {
                 let (accessor, field_type) =
                     get_ff_reader(reader, field_name, Some(get_numeric_or_date_column_types()))?;
 
-                (accessor, field_type)
+                vec![(accessor, field_type)]
             }
             Percentiles(percentiles) => {
                 let (accessor, field_type) = get_ff_reader(
                     reader,
                     percentiles.field_name(),
                     Some(get_numeric_or_date_column_types()),
                 )?;
-                (accessor, field_type)
+                vec![(accessor, field_type)]
             }
         };
 
-        let sub_aggregation = sub_aggregation.clone();
-        Ok(AggregationWithAccessor {
-            accessor,
-            accessor2,
-            field_type,
-            sub_aggregation: get_aggs_with_segment_accessor_and_validate(
-                &sub_aggregation,
-                reader,
-                &limits,
-            )?,
-            agg: agg.clone(),
-            str_dict_column,
-            limits: limits.new_guard(),
-            column_block_accessor: Default::default(),
-        })
+        let aggs: Vec<_> = acc_field_types
+            .into_iter()
+            .map(|(accessor, field_type)| {
+                Ok(AggregationWithAccessor {
+                    accessor,
+                    field_type,
+                    sub_aggregation: get_aggs_with_segment_accessor_and_validate(
+                        sub_aggregation,
+                        reader,
+                        &limits,
+                    )?,
+                    agg: agg.clone(),
+                    str_dict_column: str_dict_column.clone(),
+                    limits: limits.new_guard(),
+                    column_block_accessor: Default::default(),
+                })
+            })
+            .collect::<crate::Result<_>>()?;
```

Comment:
I think the return type fixes the types here, so that `let aggs: ...`, `Ok(aggs)` and the `?` are unnecessary, i.e. it could just end in `}).collect()`.

Reply:
I don't understand, we need to collect to Result to forward the error

Reply:
Certainly, but the extra binding and type hints seem unnecessary because the type that is collected already matches the return type, does it not?

So in think the function could just end in

```rust
        acc_field_types
            .into_iter()
            .map(|(accessor, field_type)| {
                Ok(AggregationWithAccessor {
                    accessor,
                    field_type,
                    sub_aggregation: get_aggs_with_segment_accessor_and_validate(
                        sub_aggregation,
                        reader,
                        &limits,
                    )?,
                    agg: agg.clone(),
                    str_dict_column: str_dict_column.clone(),
                    limits: limits.new_guard(),
                    column_block_accessor: Default::default(),
                })
            })
            .collect()
```

Reply:
we return a `Result<Vec>` not `Vec<Result>`

Reply:
Yes, which is what you are collecting by giving the type hint `::<crate::Result<_>>` but that type hint just matches the return type. I am not suggesting to change any types, just to drop the unnecessary temporary binding and error forwarding because the collection already produces exactly the required type.

(Turning `Iterator<Item=Result<S>>` into `Result<T> where T: FromIterator<Item=S>` is the usual [`FromIterator` impl](https://doc.rust-lang.org/stable/std/iter/trait.FromIterator.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E) which you invoke either way.)

Reply:
I think the binding with type assignment increases readability

Reply:
I don't agree, the binding and type serves the purpose of increased readability by giving it a name and a type, which is important if the transformation gets nested

Reply:
I agree with @PSeitz . It is more readable when we put the type on the variable side.